src/util/gifRecorder.ts

```diff
@@ -0,0 +1,76 @@
+import { Recorder } from './recorder';
+import gifEncoder = require('gif-encoder-2-browser');
+import { createCanvas } from 'canvas';
+
+interface GifRecordingContext {
+  interval: NodeJS.Timer;
+  frames: ImageData[];
+  height: number;
+  width: number;
+  frameRate: number;
+  onDataReady: (data: Blob) => void;
+}
+
+export class GifRecorder extends Recorder<Blob, GifRecordingContext> {
+  protected readonly format = 'GIF';
+  constructor(private readonly canvas: HTMLCanvasElement) {
+    super();
+  }
+
+  private recordingContext: GifRecordingContext | undefined;
+
+  protected startRecordingImplementation(onDataReady: (data: Blob) => void): GifRecordingContext {
+    const canvas = this.canvas;
+    const height = canvas.height;
+    const width = canvas.width;
+    const frameRate = 80;
+
+    this.recordFramesGifRecording();
+    const interval = setInterval(() => this.recordFramesGifRecording(), frameRate);
+
+    this.recordingContext = {
+      interval: interval,
+      frames: [],
+      height: height,
+      width: width,
+      frameRate: frameRate,
+      onDataReady: onDataReady,
+    };
+
+    return this.recordingContext;
+  }
+
+  private recordFramesGifRecording(): void {
+    const context = this.canvas?.getContext('2d');
+    if (context && this.recordingContext) {
+      this.recordingContext.frames.push(context.getImageData(0, 0, this.recordingContext.width, this.recordingContext.height));
+    }
+  }
```

Comment:
Move this logic to the `setInterval` callback and use the variables in the local scope (`frames`, `width`, `height`).